Cesium.Runtime/StringFunctions.cs

```diff
@@ -14,26 +7,22 @@ public unsafe static class StringFunctions
 {
     public static nuint StrLen(byte* str)
     {
-#if NETSTANDARD
-        if (str == null)
+        if (str != null)
         {
-            return 0;
-        }
+            int match;
+            nuint offset = 0;
 
-        Encoding encoding = Encoding.UTF8;
-        int byteLength = 0;
-        byte* search = str;
-        while (*search != '\0')
-        {
-            byteLength++;
-            search++;
+            // TODO: Copy IndexOfNullByte impl. from CoreLib in a distant future
+            while ((match = new Span<byte>(str + offset, int.MaxValue)
+                .IndexOf((byte)0)) < 0)
```

Comment:
While I value the performance of these vectorized functions, the use of `int.MaxValue` here seems hazardous.

In particular, there seems to not be any guarantee of `Span.IndexOfAny` not accessing the bytes after the needle is found. See the implementation: it will happily load data in chunks of 8 or 4 bytes.

If said data lies at the end of the memory page, this will easily trigger a memory access violation.

I was able to reproduce some sort of runtime crash in the following test:
```csharp
[Fact]
public void StrLen2()
{
    for (var i = 1_000_000; i < 2_000_000; ++i)
    {
        var init = new string('x', i);
        var bytes = Encoding.UTF8.GetBytes(init).Append((byte)0).ToArray();
        fixed (byte* str = bytes)
        {
            var actual = StringFunctions.StrLen(str);

            Assert.Equal((nuint)i, actual);
        }
    }
}
```

Not sure how to proceed from here, though. What do you think? To me, it looks like it's _impossible in principle_ to properly determine the string length in a vectorized way, because we don't know how long ahead we are allowed to read.

We can perform this in `StrChr` if we determine `StrLen` first, and then perform the vectorized search. But I'm not sure how efficient it will be practically. We can try to benchmark this, I guess.

But I think I'd prefer a dumb but correct implementation first.

Reply:
> Not sure how to proceed from here, though. What do you think? To me, it looks like it's impossible in principle to properly determine the string length in a vectorized way, because we don't know how long ahead we are allowed to read.

We have three options here:
- give up and revert
- calculate the distance from the pointer to memory page end (cheap), and then scan in pagesize-long chunks
- properly adapt IndexOf... code from SpanHelpers in CoreLib and write a vectorized routine here that would do the right thing and won't spend time/overhead on calling IndexOf (I'm yet to check what other strchr implementations do and if they are not vectorized manually but auto-vectorized by compiler, what the asm looks like)

I'm gonna give it a try and if it's too much code, will put it in a gist or comment out and revert until Cesium is more feature-complete :)

Reply:
Hey :) how it can become more feature complete (and what this is mean in this context?) if people like you back off of improvements ;)

also if you want very specific feature from Cesium, please tell. We would try to make it happens

Reply:
> calculate the page boundary offset, scan once with `new Span<byte>(ptr, pageEnd)`, and then scan in pagesize-long chunks

I am against the page-boundary shenanigans, just because there are no guarantees of memory pages even existing in .NET runtime. How will WASM or an alternate runtime work with them? No idea. So, we'll need a fallback solution for this case, and if we already have a fallback, then I'm okay with just using it always, to reduce testing surface.

> properly adapt IndexOf... code from SpanHelpers in CoreLib and write a vectorized routine here that would do the right thing

My point is that "the right thing" is _impossible_ to write in a vectorized way _before_ determining the boundaries of the memory region we operate on. So, vectorized `strlen` is impossible.

(Also I wonder how's it done in other C implementations.)

Reply:
> (Also I wonder how's it done in other C implementations.)

Those do it with aligning the pointer and then doing register-sized loads and bitmasks: https://elixir.bootlin.com/glibc/glibc-2.17/source/string/strlen.c

I assume page-boundary shenanigans are easily done with calculating the distance from the pointer to the next 4KiB-boundary or whatever pagesize is if we get it from the system.

Reply:
For Cesium, .NET is the target platform, and ideally it should not leak nor rely on any underlying details.

Of course, this mindset may not always be practical. But for this particular case, I believe it is still.

Reply:
We're relying on the details of .NET itself which has a few expectations from the platforms it supports like requiring support for unaligned reads/writes, specific FP math and its exceptions handling, etc. As long as we fulfill the behavior contract - it can be considered a correct implementation.

I think there is little profit in debating this at the current stage of development of Cesium and can be revisited later.

Reply:
Verified the aforementioned test on the new code, and it works now.